src/advuilist_sourced.h

```diff
@@ -0,0 +1,351 @@
+        void _registerSrc( slotidx_t c );
+        void _ctxthandler( advuilist<Container, T> *ui, std::string const &action );
+        void _printmap();
+        void _cycleslot( slotidx_t idx );
+        std::size_t _countactive( slotidx_t idx );
+
+        // used only for source map window
+        static constexpr int const _headersize = 1;
+        static constexpr int const _footersize = 1;
+        static constexpr int const _firstcol = 1;
+        static constexpr int const _iconwidth = 3;
+};
+
+// *INDENT-OFF*
+template <class Container, typename T>
+advuilist_sourced<Container, T>::advuilist_sourced( point const &srclayout, point size,
+                                                    point origin, std::string const &ctxtname )
+    : advuilist<Container, T>( &_container, size, origin, ctxtname, false ),
+      _size( size.x > 0 ? size.x : TERMX / 2, size.y > 0 ? size.y : TERMY ),
+      _origin( origin.x >= 0 ? origin.x : TERMX / 2 - _size.x / 2, origin.y >= 0 ? origin.y : 0 ),
+      _map_size( srclayout )
+// *INDENT-ON*
+
+{
+    using namespace advuilist_literals;
+    // leave room for source map window
+    point const offset( 0, _headersize + _footersize + _map_size.y );
+    advuilist<Container, T>::_resize( _size - offset, _origin + offset );
+
+    advuilist<Container, T>::_init();
+
+    advuilist<Container, T>::setctxthandler(
+        [this]( advuilist<Container, T> *ui, std::string const & action )
+    {
+        advuilist_sourced<Container, T>::_ctxthandler( ui, action );
+    } );
+
+    advuilist<Container, T>::get_ctxt()->register_action( ACTION_CYCLE_SOURCES );
+    advuilist<Container, T>::get_ctxt()->register_action( ACTION_NEXT_SLOT );
+    advuilist<Container, T>::get_ctxt()->register_action( ACTION_PREV_SLOT );
+}
+
+template <class Container, typename T>
+void advuilist_sourced<Container, T>::addSource( slotidx_t slot, source_t const &src )
+{
+    icon_t const icon = std::get<icon_t>( src );
+    auto it = _sources.find( slot );
+    if( it != _sources.end() ) {
+        std::get<slotcont_t>( it->second )[icon] = src;
+    } else {
+        // FIXME: figure out emplace for this
+        _sources[slot] = slot_t( icon, slotcont_t() );
+        std::get<slotcont_t>( _sources[slot] )[icon] = src;
+
+        _registerSrc( slot );
+    }
+}
+
+// FIXME: try to clean this up on a fresh day
+template <class Container, typename T>
+void advuilist_sourced<Container, T>::setSource( slotidx_t slot, icon_t icon, bool fallthrough )
+{
+    slot_t &_slot = _sources[slot];
+    slotcont_t &slotcont = std::get<slotcont_t>( _slot );
+    icon_t _icon = icon == 0 ? std::get<icon_t>( _slot ) : icon;
+    // only set the icon if it (still) exists in this slot, otherwise use the first one in slot
+    // rebuild() needs to additionally check that the source is available
+    std::size_t const exists = slotcont.count( _icon );
+    if( exists == 0 or ( fallthrough and !std::get<fsourceb_t>( slotcont[_icon] )() ) ) {
```

Comment:
Please do not use alternative tokens for logical operators
```suggestion
    if( exists == 0 or ( fallthrough && !std::get<fsourceb_t>( slotcont[_icon] )() ) ) {
```